crates/nu-std/lib/iter.nu

```diff
@@ -0,0 +1,113 @@
+# | Filter Extensions
+# 
+# This module implements extensions to the `filters` commands
+#
+# They are prefixed with `iter` so as to avoid conflicts with 
+# the inbuilt filters
+
+# Returns the first element of the list that matches the
+# closure predicate, `null` otherwise
+#
+# # Invariant
+# > The closure has to be a predicate (returning a bool value)
+# > else `null` is returned
+# > The closure also has to be valid for the types it receives
+# > These will be flagged as errors later as closure annotations
+# > are implemented
+#
+# # Example
+# ```
+# use std ["assert equal" "iter find"]
+#
+# let haystack = ["shell", "abc", "around", "nushell", "std"] 
+#
+# let found = ($haystack | iter find {|it| $it starts-with "a" })
+# let not_found = ($haystack | iter find {|it| $it mod 2 == 0})
+# 
+# assert equal $found "abc"
+# assert equal $not_found null
+# ```
+export def "iter find" [ # -> any | null  
+    predicate: closure   # the closure used to perform the search 
+] {
+    let list = (self collect)
+    try {
+       $list | filter $predicate | get 0?
+    } catch {
+       null
+    }
+}
+
+# Returns a new list with the separator between adjacent
+# items of the original list
+#
+# # Example
+# ```
+# use std ["assert equal" "iter intersperse"]
+#
+# let res = ([1 2 3 4] | iter intersperse 0)
+# assert equal $res [1 0 2 0 3 0 4]
+# ```
+export def "iter intersperse" [ # -> list<any>
+    separator: any,             # the separator to be used
+] {
+    let list = (self collect)
+
+    let len = ($list | length);
+    if ($list | is-empty) {
+        return $list
+    }
+
+    $list 
+    | enumerate
+    | reduce -f [] {|it, acc|
+        if ($it.index == $len - 1) {
+           $acc ++ [$it.item]
+        } else {
+            $acc ++ [$it.item, $separator]
+        }
+    }
+}
+
+# Returns the a list of intermediate steps performed by `reduce`
+# (`fold`). It takes two arguments, an initial value to seed the
+# initial state and a closure that takes two arguments, the first
+# being the internal state and the second the list element in the
+# current iteration.
+#
+# # Example
+# ```
+# use std ["assert equal" "iter scan"]
+# let scanned = ([1 2 3] | scan 0 {|x, y| $x + $y})
+#
+# assert equal $scanned [0, 1, 3, 6]
+#
+# # use the --noinit(-n) flag to remove the initial value from
+# # the final result
+# let scanned = ([1 2 3] | scan 0 {|x, y| $x + $y} -n)
+#
+# assert equal $scanned [1, 3, 6]
+# ```
+export def "iter scan" [ # -> list<any>
```

Comment:
i'm not sure i would have guessed what this command does without the examples :thinking: 
what about `iter accumulate`? :yum:

Reply:
I vote for common names.

Reply:
I think good names are the hardest part of programming, so pardon me for belaboring a point here.

I'm OK with Nu inviting analogies with Rust, especially if that's the primary language that it borrows from.  So I'd eventually get calling this function `scan` (though I'd be annoyed if it turns out to work subtly differently in Nushell).  `Accumulate` is a term used in Rust too, though not for a verb in the API, so I'd be less happy about that choice.

But roping Haskell into this discussion seems to be a stretch.  I don't know Haskell and all I can find with a quick search is a Haskell `scanl` and a `scanr`, that doesn't clarify anything for me.

Reply:
@1Kinoti
thanks for the explaination, i just did not know it was a common name :wink: 

that's perfectly fine with me then :relieved: